x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/tokenizers/BasicTokenizer.java

```diff
@@ -0,0 +1,273 @@
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+package org.elasticsearch.xpack.ml.inference.tokenizers;
+
+import joptsimple.internal.Strings;
+
+import java.text.Normalizer;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Locale;
+import java.util.Set;
+
+/**
+ * Basic tokenization of text by whitespace with optional extras:
+ * 1. Lower case the input
+ * 2. Convert to Unicode NFD
+ * 3. Stip accents
+ * 4. Surround CJK characters with ' '
+ *
+ * Derived from
+ * https://github.com/huggingface/transformers/blob/ba8c4d0ac04acfcdbdeaed954f698d6d5ec3e532/src/transformers/tokenization_bert.py
+ */
+public class BasicTokenizer {
+
+    private final boolean isLowerCase;
+    private final boolean isTokenizeCjkChars;
+    private final boolean isStripAccents;
+    private final Set<String> neverSplit;
+
+    /**
+     * Tokenizer behaviour is controlled by the options passed here.
+     *
+     * @param isLowerCase  If true convert the input to lowercase
+     * @param isTokenizeCjkChars Should CJK ideographs be tokenized
+     * @param isStripAccents Strip all accents
+     * @param neverSplit The set of tokens that should not be split
+     */
+    public BasicTokenizer(boolean isLowerCase, boolean isTokenizeCjkChars, boolean isStripAccents,
+                          Set<String> neverSplit) {
+        this.isLowerCase = isLowerCase;
+        this.isTokenizeCjkChars = isTokenizeCjkChars;
+        this.isStripAccents = isStripAccents;
+        this.neverSplit = neverSplit;
+    }
+
+    public BasicTokenizer(boolean isLowerCase, boolean isTokenizeCjkChars, boolean isStripAccents) {
+        this.isLowerCase = isLowerCase;
+        this.isTokenizeCjkChars = isTokenizeCjkChars;
+        this.isStripAccents = isStripAccents;
+        this.neverSplit = Collections.emptySet();
+    }
+
+    public BasicTokenizer(boolean isLowerCase, boolean isTokenizeCjkChars) {
+        this(isLowerCase, isTokenizeCjkChars, isLowerCase);
+    }
+
+    public BasicTokenizer() {
+        this(true, true, true);
+    }
+
+    /**
+     * Clean the text and whitespace tokenize then process depending
+     * on the values of {@code lowerCase}, {@code tokenizeCjkChars},
+     * {@code stripAccents} and the contents of {@code neverSplit}
+     *
+     * @param text The input text to tokenize
+     * @return List of tokens
+     */
+    public List<String> tokenize(String text) {
+        text = cleanText(text);
+        if (isTokenizeCjkChars) {
+            text = tokenizeCjkChars(text);
+        }
+
+        String [] tokens = whiteSpaceTokenize(text);
+
+        List<String> processedTokens = new ArrayList<>(tokens.length);
+        for (String token : tokens) {
+            if (neverSplit.contains(token)) {
+                processedTokens.add(token);
+                continue;
+            }
```

Comment:
My understanding in pseudo code is 

- If token is in never_split then skip split and lowercase logic

The `_run_split_on_punc` method has the logic

```
    def _run_split_on_punc(self, text, never_split=None):       
        if never_split is not None and text in never_split:
            return [text]
...
```

- If token is in never_split then skip split on punct

This is actually required because otherwise the special symbols `[SEP]`, `[CLS]` etc will be split on the brackets

Is the logic in this loop not the same?


Also this extra logic is required because the Python `str.split()` function removes empty strings but Java `String.split(" ")` does not

```
            if (Strings.EMPTY.equals(token)) {
                continue;
            }
```

Reply:
ugh, after reading it another 5 times, sleeping, and drinking coffee, You are correct.

My bad :D